engine/consensus/approvals/assignment_collector_statemachine_test.go

```diff
@@ -0,0 +1,107 @@
+	"testing"
+	"time"
+
+	"github.com/gammazero/workerpool"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// AssignmentCollectorStateMachineTestSuite is a test suite for testing AssignmentCollectorStateMachine. Contains a minimal set of
+// helper mocks to test the behavior.
+type AssignmentCollectorStateMachineTestSuite struct {
+	BaseAssignmentCollectorTestSuite
+	collector *AssignmentCollectorStateMachine
+}
+
+func TestAssignmentCollectorStateMachine(t *testing.T) {
+	suite.Run(t, new(AssignmentCollectorStateMachineTestSuite))
+}
+
+func (s *AssignmentCollectorStateMachineTestSuite) SetupTest() {
+	s.BaseAssignmentCollectorTestSuite.SetupTest()
+
+	s.collector = NewAssignmentCollectorStateMachine(AssignmentCollectorBase{
+		base: base{
+			workerPool:                           workerpool.New(4),
+			assigner:                             s.assigner,
+			state:                                s.state,
+			headers:                              s.headers,
+			verifier:                             s.sigVerifier,
+			seals:                                s.sealsPL,
+			approvalConduit:                      s.conduit,
+			requestTracker:                       s.requestTracker,
+			requiredApprovalsForSealConstruction: 5,
+		},
+		executedBlock: &s.Block,
+		result:        s.IncorporatedResult.Result,
+		resultID:      s.IncorporatedResult.Result.ID(),
+	})
+}
+
+// TestChangeProcessingStatus_CachingToVerifying tests that state machine correctly performs transition from CachingApprovals to
+// VerifyingApprovals state. After transition all caches approvals and results need to be applied to new state.
+func (s *AssignmentCollectorStateMachineTestSuite) TestChangeProcessingStatus_CachingToVerifying() {
+	require.Equal(s.T(), CachingApprovals, s.collector.ProcessingStatus())
+	results := make([]*flow.IncorporatedResult, 0, 3)
+	approvals := make([]*flow.ResultApproval, 0, s.Chunks.Len())
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		for i := range results {
+			block := unittest.BlockHeaderWithParentFixture(&s.Block)
+			s.blocks[block.ID()] = &block
+			result := unittest.IncorporatedResult.Fixture(
+				unittest.IncorporatedResult.WithIncorporatedBlockID(block.ID()),
+				unittest.IncorporatedResult.WithResult(s.IncorporatedResult.Result),
+			)
+			results[i] = result
+			require.NoError(s.T(), s.collector.ProcessIncorporatedResult(result))
+		}
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		for i := range approvals {
+			approval := unittest.ResultApprovalFixture(
+				unittest.WithExecutionResultID(s.IncorporatedResult.Result.ID()),
+				unittest.WithChunk(uint64(i)),
+				unittest.WithBlockID(s.Block.ID()),
+			)
+			approvals[i] = approval
+			require.NoError(s.T(), s.collector.ProcessApproval(approval))
+		}
+		wg.Done()
+	}()
+
+	err := s.collector.ChangeProcessingStatus(CachingApprovals, VerifyingApprovals)
```

Comment:
Very nice tests. 

I think this is called concurrently with the ProcessApproval and ProcessIncorporatedResult.

I wonder if you have tried this tests is able to capture the concurrency issue by [re-processing the data](https://github.com/onflow/flow-go/pull/776/files#diff-296a0ac12e421197eccaa1ebe8418d7212c2f1c875fe5ed38fbe9d15f05bea8fR63-R65)

Did you try removing [that logic](https://github.com/onflow/flow-go/pull/776/files#diff-296a0ac12e421197eccaa1ebe8418d7212c2f1c875fe5ed38fbe9d15f05bea8fR63-R65), and see if the tests fails, and added back would make the tests pass?

This could prove we did reproduce the concurrency issue, and did address it.

Reply:
Thank you, I've tried removing logic to prove that it works. 
There was some issues in the test itself so I have fixed those.